### PR TITLE
Fix a handful of OS X test failures around KVO, KVC, and Keyed Unarchival.

### LINKED ISSUE #702: Archival.NSKeyedUnarchiver_Secure fails on OSX
Dependent on https://github.com/Microsoft/WinObjC/pull/634

[ RUN      ] Archival.NSKeyedUnarchiver_Secure
Foundation/ArchivalTests.mm:138: Failure
Expected: secureUnarchiver.requiresSecureCoding = __objc_no throws an exception.
  Actual: it doesn't.
It should be impossible to turn off secure coding.
Foundation/ArchivalTests.mm:262: Failure
Value of: [secureUnarchiver decodeObjectOfClass:[NSDictionary class] forKey:@"dictionary"]
  Actual: 0x7ff2e1d24300
Expected: nullptr
Which is: NULL
Foundation/ArchivalTests.mm:263: Failure
Value of: [secureUnarchiver decodeObjectOfClass:[NSArray class] forKey:@"array"]
  Actual: 0x7ff2e1c415b0
Expected: nullptr
Which is: NULL
Foundation/ArchivalTests.mm:264: Failure
Value of: [secureUnarchiver decodeObjectOfClass:[NSNumber class] forKey:@"number"]
  Actual: 0x2837
Expected: nullptr
Which is: NULL
[  FAILED  ] Archival.NSKeyedUnarchiver_Secure (5 ms)

Not sure about the exception throwing, but the latter failures look like us having the encode/decode keys differently from OSX - not sure if that's fixable...

### LINKED ISSUE #703: KVO tests fail on OSX
Dependent on https://github.com/Microsoft/WinObjC/pull/634

[ RUN      ] KVO.DerivedKeyOnSubpath1
unknown file: Failure
Unknown C++ exception thrown in the test body.
[ RUN      ] KVO.RemoveWithoutContext
unknown file: Failure
Unknown C++ exception thrown in the test body.

Require deeper investigation to see what the exceptions actually are.

### LINKED ISSUE #755: NSKeyValueCoding tests fail on OSX
Also if we could Aggerate -> Aggregate it'd be nice

[ RUN      ] NSKeyValueCoding.AggerateFunctionUnionOfObjectsSet
unknown file: Failure
Unknown C++ exception thrown in the test body.
[  FAILED  ] NSKeyValueCoding.AggerateFunctionUnionOfObjectsSet (0 ms)
[ RUN      ] NSKeyValueCoding.AggerateFunctionSumDict
unknown file: Failure
Unknown C++ exception thrown in the test body.
[  FAILED  ] NSKeyValueCoding.AggerateFunctionSumDict (1 ms)
[ RUN      ] NSKeyValueCoding.AggerateFunctionCountDict
[       OK ] NSKeyValueCoding.AggerateFunctionCountDict (0 ms)
[ RUN      ] NSKeyValueCoding.AggerateFunctionMaxDict
unknown file: Failure
Unknown C++ exception thrown in the test body.
[  FAILED  ] NSKeyValueCoding.AggerateFunctionMaxDict (0 ms)
[ RUN      ] NSKeyValueCoding.AggerateFunctionMinDict
unknown file: Failure
Unknown C++ exception thrown in the test body.
[  FAILED  ] NSKeyValueCoding.AggerateFunctionMinDict (0 ms)
[ RUN      ] NSKeyValueCoding.AggerateFunctionAvgDict
unknown file: Failure
Unknown C++ exception thrown in the test body.
[  FAILED  ] NSKeyValueCoding.AggerateFunctionAvgDict (0 ms)
[ RUN      ] NSKeyValueCoding.AggerateFunctionDistinctUnionOfObjectsDict
unknown file: Failure
Unknown C++ exception thrown in the test body.
[  FAILED  ] NSKeyValueCoding.AggerateFunctionDistinctUnionOfObjectsDict (0 ms)
[ RUN      ] NSKeyValueCoding.AggerateFunctionUnionOfObjectsDict
[       OK ] NSKeyValueCoding.AggerateFunctionUnionOfObjectsDict (0 ms)
[ RUN      ] NSKeyValueCoding.AggerateFunctionNullLeafArray
Foundation/NSKeyValueCodingTests.mm:228: Failure
Expected: [result valueForKeyPath:@"@unionOfArrays.testObj"] throws an exception.
  Actual: it doesn't.
[  FAILED  ] NSKeyValueCoding.AggerateFunctionNullLeafArray (0 ms)
[ RUN      ] NSKeyValueCoding.AggerateFunctionNullLeafSet
Foundation/NSKeyValueCodingTests.mm:235: Failure
Expected: [result valueForKeyPath:@"@distinctUnionOfSets.testObj"] throws an exception.
  Actual: it doesn't.
[  FAILED  ] NSKeyValueCoding.AggerateFunctionNullLeafSet (0 ms)

### PR DESCRIPTION
**Pull request by DHowett-MSFT**
* Fix our illegal implementation of an observed KVO property.

    Since KVO relies heavily on tracking object state, it is illegal for an
    observed property getter to return a new object every time it is called.
    This breaks the accounting for those objects, as the first result will
    be observed, the second will be modified, the third will attempt
    deregistration, and so on.

    Fixes #703.

* Split NSSecureCoding tests, disable ones where OS X does not comply with docs. Fixes #702.
* Fix #755 by disabling 50% of our KVC aggregate tests. (Tracked in #800.)
* Fix the KVO removal-without-context test to not be implementation-dependent.

    Previously, it would expect that a widlcard context remove would take
    out the first observer, leaving the second. This is not a valid
    assumption, as the implementation is free to do as it wishes.

    Additionally, this commit wraps some statements that should absolutely
    not fail in EXPECT_NO_THROW so that we get proper test logging instead
    of generic exception handling.

    Refer to #703.

### CODE REVIEW

**Comment by DHowett-MSFT** [comment]
The best files view for this pull request is [the one without whitespace differences](https://github.com/Microsoft/WinObjC/pull/809/files?w=1).

File: tests/unittests/Foundation/FoundationTests.m [hunk 1]
```diff
@@ -286,11 +297,18 @@ + (BOOL)automaticallyNotifiesObserversOfNonNotifyingObjectProperty {
...
 - (NSString*)derivedObjectProperty {
-    return [NSString stringWithFormat:@"!!!%@!!!", _basicObjectProperty];
+    return _internal_derivedObjectProperty;
+}
+
+- (void)setBasicObjectProperty:(NSString*)basicObjectProperty {
+    [_basicObjectProperty release];
+    _basicObjectProperty = [basicObjectProperty retain];
+    _internal_derivedObjectProperty = [[NSString stringWithFormat:@"!!!%@!!!", _basicObjectProperty] retain];
+    _internal_keyDerivedTwoTimes = [[NSString stringWithFormat:@"---%@---", [self derivedObjectProperty]] retain];
```
Comment (DHowett-MSFT) [inline tests/unittests/Foundation/FoundationTests.m]: >_internal_keyDerivedTwoTimes [](start = 4, length = 28)

Please release these, or use StrongId.

**Comment by bbowman** [comment]
:shipit:

**Comment by MSFTFox** [comment]
:shipit:

**Comment by DHowett-MSFT** [comment]
Passed CI build. Prepared to merge, but I want a little more oversight :)

**Comment by msft-Jeyaram** [comment]
:shipit:

**Comment by rajsesh-msft** [comment]
:shipit:

**Comment by DHowett-MSFT** [comment]
Thank you!

**Pull request merged**